Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp

```diff
@@ -0,0 +1,132 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebCodecsAudioDataAlgorithms.h"
+
+#if ENABLE(WEB_CODECS)
+
+#include "ExceptionCode.h"
+#include "ExceptionOr.h"
+
+namespace WebCore {
+
+// https://www.w3.org/TR/webcodecs/#valid-audiodatainit
+bool isValidAudioDataInit(const WebCodecsAudioData::Init& init)
+{
+    if (init.sampleRate <= 0)
+        return false;
+
+    if (!init.numberOfFrames)
+        return false;
+
+    if (!init.numberOfChannels)
+        return false;
+
+    auto totalSamples = init.numberOfFrames * init.numberOfChannels;
+    auto bytesPerSample = computeBytesPerSample(init.format);
+    auto totalSize = totalSamples * bytesPerSample;
+    auto dataSize = init.data.span().size();
+    return dataSize >= totalSize;
+}
+
+bool isAudioSampleFormatInterleaved(const AudioSampleFormat& format)
+{
+    switch (format) {
+    case AudioSampleFormat::U8:
+    case AudioSampleFormat::S16:
+    case AudioSampleFormat::S32:
+    case AudioSampleFormat::F32:
+        return true;
+    case AudioSampleFormat::U8Planar:
+    case AudioSampleFormat::S16Planar:
+    case AudioSampleFormat::S32Planar:
+    case AudioSampleFormat::F32Planar:
+        return false;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
+    return false;
+}
+
+size_t computeBytesPerSample(const AudioSampleFormat& format)
+{
+    switch (format) {
+    case AudioSampleFormat::U8:
+    case AudioSampleFormat::U8Planar:
+        return 1;
+    case AudioSampleFormat::S16:
+    case AudioSampleFormat::S16Planar:
+        return 2;
+    case AudioSampleFormat::S32:
+    case AudioSampleFormat::F32:
+    case AudioSampleFormat::S32Planar:
+    case AudioSampleFormat::F32Planar:
+        return 4;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
+    return 0;
+}
+
+// https://www.w3.org/TR/webcodecs/#compute-copy-element-count
+ExceptionOr<size_t> computeCopyElementCount(const WebCodecsAudioData& data, const WebCodecsAudioData::CopyToOptions& options)
+{
+    auto destFormat = options.format.value_or(*data.format());
```

Comment:
How do we know data.format() is not null opt here?

Reply:
This is ambiguous indeed, I'll refactor this a bit :)